# Laboratorio 2: Limpieza de datos

# Librerias

In [1]:
'''NumPy es una librería para el cálculo numérico y el análisis de datos'''
import numpy as np
'''Pandas para el análisis y manipulación de datos'''
import pandas as pd
'''El módulo csv para leer y escribir datos tabulares en formato CSV'''
import csv

# Lectura del Dataset Criminología

In [2]:
'''Carga del dataset Criminology mediante la libreía Pandas '''
df = pd.read_csv('Criminology.csv', sep=';')
'''Lectura del encabezado del dataset Criminology'''
df

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
0      160903280     Assault / Battery  2016-03-30T00:00:00    18:42   
1      160912272    Homeless Complaint  2016-03-31T00:00:00    15:31   
2      160912590             Susp Info  2016-03-31T00:00:00    16:49   
3      160912801                Report  2016-03-31T00:00:00    17:38   
4      160912811                   594  2016-03-31T00:00:00    17:42   
...          ...                   ...                  ...      ...   
10046  160964210      Well Being Check  2016-04-05T00:00:00    23:38   
10047  160964216          Traffic Stop  2016-04-05T00:00:00    23:43   
10048  160964227          Traffic Stop  2016-04-05T00:00:00    23:48   
10049  160964229          Passing Call  2016-04-05T00:00:00    23:48   
10050  160964249          Traffic Stop  2016-04-05T00:00:00    23:54   

              CallDateTime   Disposition                  Address  \
0      2016-03-30T18:42:00           REP  100 Block Of Chilton Av   
1      2016-03-31T15:31:00           GOA  2300 Block Of Market St   
2      2016-03-31T16:49:00           GOA  2300 Block Of Market St   
3      2016-03-31T17:38:00           GOA      500 Block Of 7th St   
4      2016-03-31T17:42:00           REP       Beale St/bryant St   
...                    ...           ...                      ...   
10046  2016-04-05T23:38:00           GOA    Hollister Av/hawes St   
10047  2016-04-05T23:43:00           ADV   California St/mason St   
10048  2016-04-05T23:48:00           CIT   Rosella Ct/onondaga Av   
10049  2016-04-05T23:48:00  Not recorded          0 Blk Herbst Rd   
10050  2016-04-05T23:54:00           ADV        Grant Av/geary St   

                City State AgencyId  Range      AddressType  
0      San Francisco    CA        1    NaN  Premise Address  
1      San Francisco    CA        1    NaN  Premise Address  
2      San Francisco    CA        1    NaN  Premise Address  
3      San Francisco    CA        1    NaN  Premise Address  
4      San Francisco    CA        1    NaN     Intersection  
...              ...   ...      ...    ...              ...  
10046  San Francisco    CA        1    NaN     Intersection  
10047  San Francisco    CA        1    NaN     Intersection  
10048  San Francisco    CA        1    NaN     Intersection  
10049  San Francisco    CA        1    NaN     Geo-Override  
10050  San Francisco    CA        1    NaN     Intersection  

[10051 rows x 12 columns]

## Información de cada columna
Se observa la información de cada columna con ello se procede a identificar los valores y el total de registro que posee cada columna. Alguna columnas como City, State y Range no tienen el valor total de registro que es 10051. 

In [3]:
'''Información de que tiene cada columna'''
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10051 entries, 0 to 10050
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   CrimeId                10051 non-null  int64  
 1   OriginalCrimeTypeName  10051 non-null  object 
 2   OffenseDate            10051 non-null  object 
 3   CallTime               10051 non-null  object 
 4   CallDateTime           10051 non-null  object 
 5   Disposition            10051 non-null  object 
 6   Address                10051 non-null  object 
 7   City                   9730 non-null   object 
 8   State                  10048 non-null  object 
 9   AgencyId               10051 non-null  object 
 10  Range                  0 non-null      float64
 11  AddressType            10051 non-null  object 
dtypes: float64(1), int64(1), object(10)
memory usage: 942.4+ KB


### Verificación de valores NULOS
Se procede a observar los valores nulos para tenerse en cuenta al momento de empezar a realizar la limpieza

In [4]:
'''Total de valores nulos en cada columna'''
df.isnull().sum()

CrimeId                      0
OriginalCrimeTypeName        0
OffenseDate                  0
CallTime                     0
CallDateTime                 0
Disposition                  0
Address                      0
City                       321
State                        3
AgencyId                     0
Range                    10051
AddressType                  0
dtype: int64

# Limpieza de datos
Para la limpieza de datos se procede a verificar que sean unicos. Se procede a relizar una acción con los datos que no tienen sentido para el dataset. Se va a ir examinando cada columna para verificar que esten correctas.

# Primer problema
#### Columna Range
La columna Range es una columna vacia, es decir, con valores nulos por lo tanto no es necesaria para el Dataset. Con ello se procede a ser eliminada. Esta columna ocupa espacio innecesario en el Dataset.

In [5]:
'''Eliminación de la columna Range'''
df = df.drop(['Range'],axis=1)
'''Lectura del dataset sin la columna Range'''
df

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
0      160903280     Assault / Battery  2016-03-30T00:00:00    18:42   
1      160912272    Homeless Complaint  2016-03-31T00:00:00    15:31   
2      160912590             Susp Info  2016-03-31T00:00:00    16:49   
3      160912801                Report  2016-03-31T00:00:00    17:38   
4      160912811                   594  2016-03-31T00:00:00    17:42   
...          ...                   ...                  ...      ...   
10046  160964210      Well Being Check  2016-04-05T00:00:00    23:38   
10047  160964216          Traffic Stop  2016-04-05T00:00:00    23:43   
10048  160964227          Traffic Stop  2016-04-05T00:00:00    23:48   
10049  160964229          Passing Call  2016-04-05T00:00:00    23:48   
10050  160964249          Traffic Stop  2016-04-05T00:00:00    23:54   

              CallDateTime   Disposition                  Address  \
0      2016-03-30T18:42:00           REP  100 Block Of Chilton Av   
1      2016-03-31T15:31:00           GOA  2300 Block Of Market St   
2      2016-03-31T16:49:00           GOA  2300 Block Of Market St   
3      2016-03-31T17:38:00           GOA      500 Block Of 7th St   
4      2016-03-31T17:42:00           REP       Beale St/bryant St   
...                    ...           ...                      ...   
10046  2016-04-05T23:38:00           GOA    Hollister Av/hawes St   
10047  2016-04-05T23:43:00           ADV   California St/mason St   
10048  2016-04-05T23:48:00           CIT   Rosella Ct/onondaga Av   
10049  2016-04-05T23:48:00  Not recorded          0 Blk Herbst Rd   
10050  2016-04-05T23:54:00           ADV        Grant Av/geary St   

                City State AgencyId      AddressType  
0      San Francisco    CA        1  Premise Address  
1      San Francisco    CA        1  Premise Address  
2      San Francisco    CA        1  Premise Address  
3      San Francisco    CA        1  Premise Address  
4      San Francisco    CA        1     Intersection  
...              ...   ...      ...              ...  
10046  San Francisco    CA        1     Intersection  
10047  San Francisco    CA        1     Intersection  
10048  San Francisco    CA        1     Intersection  
10049  San Francisco    CA        1     Geo-Override  
10050  San Francisco    CA        1     Intersection  

[10051 rows x 11 columns]

# Segundo Problema
#### Columna City
Se verifica que contenga valores unicos la columna City. Esta columna tiene algunos valores que no estan correctamente escritos por ello se reemplazan los campos.

In [6]:
'''Visualización de valores únicos en la columna City'''
df.City.unique()

array(['San Francisco', nan, 'Daly City', 'Treasure Isla', 'Yerba Buena',
       'SAN FRANCISCO', 'Presidio', ' S', 'Brisbane'], dtype=object)

In [7]:
'''Agrupación de registros para observar lo que hay que corregir'''
df.groupby('City').size()

City
 S                  1
Brisbane            1
Daly City           5
Presidio            3
SAN FRANCISCO       1
San Francisco    9665
Treasure Isla      51
Yerba Buena         3
dtype: int64

In [8]:
'''Reemplazo en la columna City los registros mal escritos'''
df['City'] = df['City'].replace({'SAN FRANCISCO': 'San Francisco', ' S': 'San Francisco'})
'''Verificación del reemplazo en la columna City'''
df.groupby('City').size()

City
Brisbane            1
Daly City           5
Presidio            3
San Francisco    9667
Treasure Isla      51
Yerba Buena         3
dtype: int64

#### Columna Disposition
Se muestran los valores únicos de la columna Disposition. Este muestra que tiene valores que reemplazar. Y tambien se corrige una palabra mas escrita

In [9]:
'''Visualización de valores únicos en la columna Disposition'''
df.Disposition.unique()

array(['REP', 'GOA', 'ADV', 'HAN', 'ND', 'ARR', 'UTL', 'Not recorded',
       'CIT', 'PAS', 'ABA', 'NOM', 'ADM', 'CAN', 'INC', 'NCR', '22',
       'CRT', 'SFD'], dtype=object)

In [10]:
'''Agrupación de registros para observar lo que hay que corregir'''
df.groupby('Disposition').size()

Disposition
22                77
ABA               97
ADM               48
ADV             1142
ARR               65
CAN              353
CIT             1425
CRT                2
GOA             1273
HAN             2820
INC               17
NCR               82
ND               427
NOM              324
Not recorded     543
PAS              170
REP              800
SFD                1
UTL              385
dtype: int64

In [11]:
'''Reemplazo de registros al no saber que es 22 se lo transforma a ND lo cual es No Disposition'''
df['Disposition'] = df['Disposition'].replace({'Not recorded': 'ND', '22': 'ND'})
'''Agrupación de registros para observar los reemplazos'''
df.groupby('Disposition').size()

Disposition
ABA      97
ADM      48
ADV    1142
ARR      65
CAN     353
CIT    1425
CRT       2
GOA    1273
HAN    2820
INC      17
NCR      82
ND     1047
NOM     324
PAS     170
REP     800
SFD       1
UTL     385
dtype: int64

#### Columna AddressType
Se verifica que la columna AddressType tenga valores únicos. Este muestra que tiene valores en 1 por ello se reemplaza esos valores por el valor mas repetitivo (moda = Premise Address). Tambien se corrige una palabra mas escrita

In [12]:
'''Visualización de valores únicos en la columna AddressType'''
df.AddressType.unique()

array(['Premise Address', 'Intersection', 'Common Location',
       'Geo-Override', 'Intersectioon', '1'], dtype=object)

In [13]:
'''Agrupación de registros para observar lo que hay que corregir'''
df.groupby('AddressType').size()

AddressType
1                     3
Common Location     818
Geo-Override        469
Intersection       3701
Intersectioon         1
Premise Address    5059
dtype: int64

In [14]:
'''Reemplazo del número por el registro más repetitivo'''
df['AddressType'] = df['AddressType'].replace('1','Premise Address')
df['AddressType'] = df['AddressType'].replace('Intersectioon','Intersection')

'''Verificación del reemplazo en la columna'''
df.groupby('AddressType').size()

AddressType
Common Location     818
Geo-Override        469
Intersection       3702
Premise Address    5062
dtype: int64

#### Columna OriginalCrimeTypeName
Se reempalza los registros que tienen simbología innecesaria.

In [15]:
'''Visualización de valores únicos en la columna State'''
df.OriginalCrimeTypeName.unique()

array(['Assault / Battery', 'Homeless Complaint', 'Susp Info', 'Report',
       '594', "Ref'd", 'Agg Assault / Adw Dv', 'Encampment', 'Burglary',
       'Suspicious Person', 'Ip', 'Arrest Made', 'Tent',
       'Intoxicated Person', 'Guitar', 'H&r Veh Accident', 'Music',
       'Person Dumping', 'Vandalism', 'Petty Theft', '415',
       'Noise Nuisance', 'Fraud', 'Rep', 'Meet W/citizen',
       'Prisoner Transport', 'Poss', 'Trespasser', 'Stolen Vehicle',
       'Fight No Weapon', 'Check For Ped', 'Juvenile Disturbance',
       'Party', 'Lltd', 'Robbery', '601', '500e', '913', 'Traffic Stop',
       '909', 'Person W/gun', 'Suspicious Vehicle', 'Fight No Weapon Dv',
       'Audible Alarm', 'Parking', 'Pay', 'Silent Holdup Alarm',
       'Passing Call', 'Sleeper', 'Family', 'Thumping', 'Loud Music',
       '916', 'Att', 'Threats / Harassment', 'Boombox',
       'Traf Violation Tow', 'Dw', 'Bar Check', 'Suicide Attempt', 'Susp',
       'Crosswalk', 'Casing/917', 'Wireless-Drop', 'Uc Hastin

In [16]:
'''Reemplazo de registros con simbología innecesaria'''
df['OriginalCrimeTypeName'] = df['OriginalCrimeTypeName'].replace({'*909*': '909', 
                                                                   '**Tazer**': 'Tazer',
                                                                   '***901 Req***': '901 Req',
                                                                   'Fi**expedite**': 'Fiexpedite',
                                                                   '.52': '52', '.25/500e': '25/500e', '`Drugs': 'Drug'})
'''Visualización de valores únicos en la columna State'''
df.OriginalCrimeTypeName.unique()

array(['Assault / Battery', 'Homeless Complaint', 'Susp Info', 'Report',
       '594', "Ref'd", 'Agg Assault / Adw Dv', 'Encampment', 'Burglary',
       'Suspicious Person', 'Ip', 'Arrest Made', 'Tent',
       'Intoxicated Person', 'Guitar', 'H&r Veh Accident', 'Music',
       'Person Dumping', 'Vandalism', 'Petty Theft', '415',
       'Noise Nuisance', 'Fraud', 'Rep', 'Meet W/citizen',
       'Prisoner Transport', 'Poss', 'Trespasser', 'Stolen Vehicle',
       'Fight No Weapon', 'Check For Ped', 'Juvenile Disturbance',
       'Party', 'Lltd', 'Robbery', '601', '500e', '913', 'Traffic Stop',
       '909', 'Person W/gun', 'Suspicious Vehicle', 'Fight No Weapon Dv',
       'Audible Alarm', 'Parking', 'Pay', 'Silent Holdup Alarm',
       'Passing Call', 'Sleeper', 'Family', 'Thumping', 'Loud Music',
       '916', 'Att', 'Threats / Harassment', 'Boombox',
       'Traf Violation Tow', 'Dw', 'Bar Check', 'Suicide Attempt', 'Susp',
       'Crosswalk', 'Casing/917', 'Wireless-Drop', 'Uc Hastin

# Tercer problema
#### Columna City
Se verifica que la columna City tenga valores únicos. La columna city tiene valores nulos por ello se reemplaza esos valores nulos por sin registro.

In [17]:
'''Visualización de valores únicos en la columna City'''
df.City.unique()

array(['San Francisco', nan, 'Daly City', 'Treasure Isla', 'Yerba Buena',
       'Presidio', 'Brisbane'], dtype=object)

In [18]:
'''Reemplazo de valores nulos por No Registration'''
df['City'].fillna('No Registration', inplace=True)

In [19]:
'''Se verifica que ya no haya valores nulos en la columna City'''
df.isnull().sum()

CrimeId                  0
OriginalCrimeTypeName    0
OffenseDate              0
CallTime                 0
CallDateTime             0
Disposition              0
Address                  0
City                     0
State                    3
AgencyId                 0
AddressType              0
dtype: int64

# Cuarto Problema
#### Columna State
Se verifica que la columna State sea única. Esto muestra que la columna State tiene valores nulos donde, estos son reemplazados por el campo CA

In [20]:
'''Visualización de valores únicos en la columna State'''
df.State.unique()

array(['CA', nan], dtype=object)

In [21]:
'''Agrupación de datos para saber cuantos datos existen'''
df.groupby('State').size()

State
CA    10048
dtype: int64

In [22]:
'''Se reemplaza los valores NULOS por el campo CA'''
df['State'].fillna('CA', inplace=True)

In [23]:
'''Visualización de los registros correctos en la columna State'''
df.groupby('State').size()

State
CA    10051
dtype: int64

In [24]:
'''Visualización de valores únicos en la columna State'''
df.State.unique()

array(['CA'], dtype=object)

# Quinto Problema
#### Columna Agency
Se verifica que la columna AgencyId sea única. Esta muestra que tiene campos numéricos por ende se debe convertir las letras a números ya que esos son la mayria de sus datos.

In [25]:
'''Verificación de datos únicos'''
df.AgencyId.unique()

array(['1', 'CA'], dtype=object)

In [26]:
'''Agrupación de datos para saber cuantos datos están mal registrados'''
df.groupby('AgencyId').size()

AgencyId
1     10048
CA        3
dtype: int64

In [27]:
'''Reemplazo de las letras en la columna'''
df['AgencyId'] = df['AgencyId'].replace('CA','1')
'''Verificación del reemplazo en la columna'''
df.groupby('AgencyId').size()

AgencyId
1    10051
dtype: int64

In [28]:
'''Verificación de datos únicos nuevamente'''
df.AgencyId.unique()

array(['1'], dtype=object)

# Sexto Problema
#### Columna CrimeId
Verificación de valores únicos en la columna CrimeID. Esta primer columna es el campo de registros único, es decir, no puede haber valores duplicados. Por lo tanto se revisa si contiene valores duplicados.

In [29]:
'''Visualización de valores únicos en la columna State'''
df.CrimeId.unique()

array([160903280, 160912272, 160912590, ..., 160964227, 160964229,
       160964249], dtype=int64)

##### Datos Duplicados
Por ende si existen datos duplicados no habría como establecer el campo como un indice único, por lo que se modifica los valores duplicados por valores únicos, utilizando el valor máximo existente para la columna.

In [30]:
'''Se observan los registros duplicados del dataset'''
df[df.duplicated(subset = ['CrimeId'], keep=False)]

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
26    160913455             Vandalism  2016-03-31T00:00:00    20:53   
1707  160913455                  Susp  2016-04-01T00:00:00    18:29   
3792  160913455          Passing Call  2016-04-02T00:00:00    17:11   
7045  160950496          Passing Call  2016-04-04T00:00:00     6:51   
7046  160950496    Suspicious Vehicle  2016-04-04T00:00:00     6:51   
7047  160950496            Trespasser  2016-04-04T00:00:00     6:51   

             CallDateTime Disposition                     Address  \
26    2016-03-31T20:53:00          ND  1600 Block Of Sunnydale Av   
1707  2016-04-01T18:29:00         GOA          Geary St/larkin St   
3792  2016-04-02T17:11:00          ND      900 Block Of Market St   
7045  2016-04-04T06:51:00         HAN     University St/felton St   
7046  2016-04-04T06:51:00          ND   1400 Block Of Cabrillo St   
7047  2016-04-04T06:51:00         CAN       Block Of Hampshire St   

               City State AgencyId      AddressType  
26    San Francisco    CA        1  Premise Address  
1707  San Francisco    CA        1     Intersection  
3792  San Francisco    CA        1  Premise Address  
7045  San Francisco    CA        1     Intersection  
7046  San Francisco    CA        1  Premise Address  
7047  San Francisco    CA        1  Premise Address

In [31]:
'''Se modifica los valores duplicados por valores únicos'''
duplicados = df.index[df.duplicated(subset = ['CrimeId'], keep = False)].tolist()
'''Se utiliza el valor máximo existente para la columna'''
max = df['CrimeId'].max()
'''Ciclo para establecer los valores como únicos'''
for indice in duplicados:
  max += 1 #Al valor máximo aumenta en 1
  df['CrimeId'].iloc[indice] = max #Establece el nuevo valor del indice

C:\Users\Xiomy\AppData\Local\Temp\ipykernel_4708\1731213875.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['CrimeId'].iloc[indice] = max #Establece el nuevo valor del indice


In [32]:
'''Se verfica nuevamente que no haya valores duplicados'''
df[df.duplicated(subset=['CrimeId'],keep=False)]

Empty DataFrame
Columns: [CrimeId, OriginalCrimeTypeName, OffenseDate, CallTime, CallDateTime, Disposition, Address, City, State, AgencyId, AddressType]
Index: []

# Conversión a archivo JSON
Exportación del DataFrame Limpio en archivos de formato CSV y JSON

In [33]:
'''Guardar en formato CSV el Datraframe limpio'''
df.to_csv('Criminology_Limpio.csv', sep=',', index=False, header=True)
'''Guardar en formato JASON el Datraframe limpio'''
df.to_json('Criminology_Limpio.json', orient='records', lines=True)

In [34]:
'''Lectura del conjunto de datos JSON'''
dfJSON = pd.read_json("Criminology_Limpio.json", orient='records', lines=True)
'''Lectura del Conjunto de datos JSON'''
dfJSON

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
0      160903280     Assault / Battery  2016-03-30T00:00:00    18:42   
1      160912272    Homeless Complaint  2016-03-31T00:00:00    15:31   
2      160912590             Susp Info  2016-03-31T00:00:00    16:49   
3      160912801                Report  2016-03-31T00:00:00    17:38   
4      160912811                   594  2016-03-31T00:00:00    17:42   
...          ...                   ...                  ...      ...   
10046  160964210      Well Being Check  2016-04-05T00:00:00    23:38   
10047  160964216          Traffic Stop  2016-04-05T00:00:00    23:43   
10048  160964227          Traffic Stop  2016-04-05T00:00:00    23:48   
10049  160964229          Passing Call  2016-04-05T00:00:00    23:48   
10050  160964249          Traffic Stop  2016-04-05T00:00:00    23:54   

              CallDateTime Disposition                  Address  \
0      2016-03-30T18:42:00         REP  100 Block Of Chilton Av   
1      2016-03-31T15:31:00         GOA  2300 Block Of Market St   
2      2016-03-31T16:49:00         GOA  2300 Block Of Market St   
3      2016-03-31T17:38:00         GOA      500 Block Of 7th St   
4      2016-03-31T17:42:00         REP       Beale St/bryant St   
...                    ...         ...                      ...   
10046  2016-04-05T23:38:00         GOA    Hollister Av/hawes St   
10047  2016-04-05T23:43:00         ADV   California St/mason St   
10048  2016-04-05T23:48:00         CIT   Rosella Ct/onondaga Av   
10049  2016-04-05T23:48:00          ND          0 Blk Herbst Rd   
10050  2016-04-05T23:54:00         ADV        Grant Av/geary St   

                City State  AgencyId      AddressType  
0      San Francisco    CA         1  Premise Address  
1      San Francisco    CA         1  Premise Address  
2      San Francisco    CA         1  Premise Address  
3      San Francisco    CA         1  Premise Address  
4      San Francisco    CA         1     Intersection  
...              ...   ...       ...              ...  
10046  San Francisco    CA         1     Intersection  
10047  San Francisco    CA         1     Intersection  
10048  San Francisco    CA         1     Intersection  
10049  San Francisco    CA         1     Geo-Override  
10050  San Francisco    CA         1     Intersection  

[10051 rows x 11 columns]